lib/tokenize.cpp

```diff
@@ -3563,7 +3563,7 @@ static bool setVarIdParseDeclaration(const Token **tok, const VariableMap& varia
                 tok2 = tok2->linkAt(1)->next();
                 continue;
             }
-            if (Token::Match(tok2, "struct|union|enum") || (!c && Token::Match(tok2, "class|typename"))) {
+            if (Token::Match(tok2, "struct|union|enum") || ((!c || header) && Token::Match(tok2, "class|typename"))) {
```

Comment:
I think the `|| header` is unfortunate. the question is what language it is.. and header has no information about it.

With this code a header is always assumed to contain C++ code. So if a header with C code contains some `int class;` variable declaration then after your changes that `class` variable will be treated as a C++ keyword.

If the user uses `--language=c`  then this `|| header` code will "override" that.

Reply:
That whole logic is unfortunate but this keeps the intended behavior intact - I think. At least the tests pass. I will re-visit this when my head is clearer.

Reply:
The logic is really strange. Out of the box we treat header files as C and bail out but this code wants to treat those as C++ by default. I think we should get rid of that inconsistency. I will try to perform and write a few more tests for this.

Reply:
The changed behavior appears to be consistent and should be applied IMO. I came across this again in #5853 which also exposes the flawed logic.

Reply:
The new test results are not yet final. Will take another look later today.

Reply:
Okay - I cleaned up that change a lot and updated the test results.

That looks a bit saner now overall but I am a bit baffled why we are able to parse the C++ code in C code. It should bail out and tell you that it is invalid C code.